### PR TITLE
Petsc 3.16 and Docker/Singularity info

### PR DESCRIPTION
**Pull request by piyueh**
This PR
* in `src/misc/misc.cpp`, replaces `DMDA_X`, `DMDA_Y`, `DMDA_Z` with `DM_X`, `DM_Y`, `DM_Z` (avail in PETSc since v3.13)
* in `applications/navierstokes/navierstokes.cpp`, adds a null pointer to `PetscViewerHDF5ReadAttribute` per the new signature (since PETSc v3.15)
* updates the dependency versions in `doc/markdowns/installation.md`
* updates autoconf
* adds information of Docker and Singularity images to `doc/markdowns/installation.md`
* bumps version numbers in documents to 0.5.3 for future release

_**Note**_: because the signature change in PETSc function `PetscViewerHDF5ReadAttribute` is very recent (since v3.16.0), after merging this PR, PetIBM will not be able to use any PETSc older than 3.16. And also because of using newer PETSc, I believe most older GCC compilers will not work either.

@mesnardo

### CODE REVIEW

**Comment by mesnardo** [comment]
@piyueh Looking at your PR. Will test the installation on ubuntu this evening and review the changes.

Couple questions:

* Is `DMDA_*` deprecated and replaced with `DM_*`?
* PR is currently targeting `master`. What's the plan to synchronize with `dev`? (Merge into `master`, release, and then merge into `develop`?)

**Comment by piyueh** [comment]
> @piyueh Looking at your PR. Will test the installation on ubuntu this evening and review the changes.
> 
> Couple questions:
> 
> * Is `DMDA_*` deprecated and replaced with `DM_*`?

Yes. They are not in PETSc's source code anymore since https://github.com/petsc/petsc/commit/3ee9839e4674d2e0e9bcea975330f3458ebcb61e

> * PR is currently targeting `master`. What's the plan to synchronize with `dev`? (Merge into `master`, release, and then merge into `develop`?)

I'm fine with either merging to `master` to `develop`. Actually, I completely forgot we have a `develop` branch.

Will we continue developing PetIBM? Or are we now under maintenance mode? If we will just do maintenance, then I feel like there's no need to maintain the `develop` branch. Most foreseeable PRs are probably just to make sure PetIBM runs with the latest dependencies, i.e., small patches. And if we really need the `develop` branch at some point in the future, we can then rebase `develop` onto `master`, merge new features into `develop`, and then merge to `master`. Just my two cents though.

File: doc/markdowns/installation.md [hunk 1]
```diff
@@ -28,11 +37,13 @@
 * [yaml-cpp](https://github.com/jbeder/yaml-cpp),
   [gtest](https://github.com/google/googletest), and
   [AmgXWrapper](https://github.com/barbagroup/AmgXWrapper) can be automatically installed during PetIBM configuration or explicitly installed by users in advance.
-* PetIBM has been tested on:
+* The current version of PetIBM has been tested with:
+    * Arch Linux with g++-11.2.0 and PETSc 3.16.5
```
Comment (mesnardo) [inline doc/markdowns/installation.md:41]:
```suggestion
* The latest release of PetIBM (0.5.3) has been tested with:
    * Arch Linux with g++-11.2.0 and PETSc 3.16.5
    * Ubuntu 20.04 with g++-9.4 and PETSc 3.16.5
```

File: doc/markdowns/installation.md [hunk 2]
```diff
@@ -28,11 +37,13 @@
...
+* PetIBM versions prior commit a657821 (or version 0.5.2, included) have been tested with:
```
Comment (mesnardo) [inline doc/markdowns/installation.md:42]:
```suggestion
* Previous releases of PetIBM (<=0.5.2) have at some point been tested with:
```

File: doc/markdowns/installation.md [hunk 3]
```diff
@@ -28,11 +37,13 @@
...
     * Ubuntu 16.04 with g++-5.4, and PETSc-3.11.2
     * MacOS Sierra with g++-6.0, and PETSc-3.8.2
     * Arch Linux with g++-7.2, and PETSc-3.8.2
-* PetIBM has also been tested on the following HPC systems:
+* Older versions of PetIBM have also been tested on the following HPC systems:
```
Comment (mesnardo) [inline doc/markdowns/installation.md:46]:
```suggestion
* In the past, we have also been able to run previous versions of PetIBM on the following HPC systems:
```

File: m4/configure_petsc.m4 [hunk 4]
```diff
@@ -63,7 +63,7 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM(
 #define PETSC_VERSION_GE(MAJOR, MINOR, SUBMINOR) (!PETSC_VERSION_LT(MAJOR, MINOR, SUBMINOR))
 #endif
 ]], [[
-#if ((PETSC_VERSION_GE(3, 8, 0) && PETSC_VERSION_LT(3, 13, 0)) || PETSC_VERSION_RELEASE != 1)
+#if ((PETSC_VERSION_GE(3, 16, 0) && PETSC_VERSION_LT(3, 18, 0)) || PETSC_VERSION_RELEASE != 1)
```
Comment (mesnardo) [inline m4/configure_petsc.m4:66]: I don't think the PETSc 3.17 series has started (and not sure if breaking changes will be introduced). Should we set the upper bound condition to `PETSC_VERSION_LT(3, 17, 0)`?

**Review by mesnardo — COMMENTED** [review]
Thank you @piyueh, I reviewed and left some small suggestions/questions.

I successfully tested the installation process (and running examples) inside a Docker container using:

 petibm@ 2552235, ubuntu-20.04, gcc-9.4, openmpi-3.1.4, cuda-11.6.0, petsc-3.16.5, and amgx-2.2.0

**Comment by mesnardo** [comment]
> I'm fine with either merging to master or to develop. Actually, I completely forgot we have a develop branch.

Since you are bumping the version (0.5.3) in this PR, I'm good with merging to `master`. We will do a `master`->`develop` PR to sync up after the release.

> Will we continue developing PetIBM? Or are we now under maintenance mode? If we will just do maintenance, then I feel like there's no need to maintain the develop branch. Most foreseeable PRs are probably just to make sure PetIBM runs with the latest dependencies, i.e., small patches. And if we really need the develop branch at some point in the future, we can then rebase develop onto master, merge new features into develop, and then merge to master. Just my two cents though.

Unlikely I will continue development on PetIBM in the near future. But future students of Prof. Barba may. So I would suggest to keep maintaining a `develop` branch for now (not much added work). Also, I'd rather have external feature contributions being merged into `develop` than `master`. Does that sound good to you?

File: m4/configure_petsc.m4 (hunk 4 above)
Comment (piyueh) [reply m4/configure_petsc.m4]: Done

**Comment by piyueh** [comment]
> 

Agree.

I just applied the suggested changes. Could you review them again and merge and close the PR if everything looks fine? Thanks!

**Review by mesnardo — APPROVED** [review]
LGTM

**Pull request merged**